### PR TITLE
test(http): basic endpoint tests

### LINKED ISSUE #315: todo: basic endpoint testing
Before #20 can be actioned, we must validate that route handlers mostly behave within expectations. This can be done simply by asserting that handlers respond with the expected HTTP status code, a body, basic headers, etc. I want to emphasis that this testing only needs to be basic. Some of the routes we should cover include (but aren't exclusive to):
* `GET /`
* `GET /blog`
* `GET /pricing`
* `GET /signin` and other auth pages
* `GET /account` (redirect to sign-page)

For guidance, see https://fresh.deno.dev/docs/examples/writing-tests

### PR DESCRIPTION
**Pull request by mbhrznr**
closes #315.

this pr adds a bunch of simple tests to have some coverage on our endpoints.
as described in the issue, it simply follows the guide as described [here](https://fresh.deno.dev/docs/examples/writing-tests).

additionally it tweaks the `deleteVote` function, as theoretically, one could've downvoted an item into oblivion.

### CODE REVIEW

File: utils/db.ts [hunk 1]
```diff
@@ -286,11 +279,29 @@ export async function deleteVote(vote: Vote) {
     vote.user.id,
   ];
 
+  const [votedItemsByUserRes, votedUsersByItemRes] = await kv.getMany([
```
Comment (mbhrznr) [inline utils/db.ts:282]: this is something i stumbled upon by accident.  
if we compare `createVote` and `deleteVote`, we were missing a check for the `votedItemsByUserKey` and `votedUsersByItemKey` results.

as the deletion for a non-existent key doesn't throw (anymore?), the mutation on the `vote.item.score` would've been abused via api calls.

this is why an explicit check for these keys and their respective results has been introduced.

**Comment by iuioiua** [comment]
I half expected this - test coverage has taken a significant hit now that more files are included in the tested list. What to do...

**Comment by iuioiua** [comment]
Let's ignore test coverage for `/routes`, `/components` and `/islands`. See https://docs.codecov.com/docs/ignoring-paths

**Comment by mbhrznr** [comment]
> Let's ignore test coverage for `/routes`, `/components` and `/islands`. See https://docs.codecov.com/docs/ignoring-paths

added the `ignore` section as suggested.

didn't came to mind to check the coverage, yet it makes sense that we took this big hit.  
due to `manifest` we'd import everything into the newly created http test set, correct?

as for the now ignored dirs it might make sense to introduce e2e tests at some point, wdyt?

File: utils/db.ts (hunk 1 above)
Comment (iuioiua) [reply utils/db.ts:282]: Good catch! Though, next time, let's keep the objectives per PR to 1 🙂

**Review by iuioiua — APPROVED** [review]
LGTM! Thanks for this 🙏🏾

File: tests/http_test.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+const CONN_INFO: ConnInfo = {
+  localAddr: { hostname: "localhost", port: 8000, transport: "tcp" },
+  remoteAddr: { hostname: "localhost", port: 53496, transport: "tcp" },
+};
+
+Deno.test("[http]", async (test) => {
+  const handler = await createHandler(manifest);
+
+  await test.step("[get] /", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:22]:
```suggestion
  await test.step("GET /", async () => {
```

File: tests/http_test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    assert(response.ok);
+    assertInstanceOf(response.body, ReadableStream);
+    assertEquals(
+      response.headers.get("content-type"),
+      "text/html; charset=utf-8",
+    );
+    assertEquals(response.status, 200);
+  });
+
+  await test.step("[get] /account", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:34]:
```suggestion
  await test.step("GET /account", async () => {
```

File: tests/http_test.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    assertFalse(response.ok);
+    assertFalse(response.body);
+    assertEquals(
+      response.headers.get("location"),
+      "/signin?from=http://localhost/account",
+    );
+    assertEquals(response.status, 303);
+  });
+
+  await test.step("[get] /callback", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:49]:
```suggestion
  await test.step("GET /callback", async () => {
```

File: tests/http_test.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    assertFalse(response.ok);
+    assertInstanceOf(response.body, ReadableStream);
+    assertEquals(
+      response.headers.get("content-type"),
+      "text/html; charset=utf-8",
+    );
+    assertEquals(response.status, 500);
+  });
+
+  await test.step("[get] /blog", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:64]:
```suggestion
  await test.step("GET /blog", async () => {
```

File: tests/http_test.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    assert(response.ok);
+    assertInstanceOf(response.body, ReadableStream);
+    assertEquals(
+      response.headers.get("content-type"),
+      "text/html; charset=utf-8",
+    );
+    assertEquals(response.status, 200);
+  });
+
+  await test.step("[get] /pricing", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:79]:
```suggestion
  await test.step("GET /pricing", async () => {
```

File: tests/http_test.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    assertFalse(response.ok);
+    assertInstanceOf(response.body, ReadableStream);
+    assertEquals(
+      response.headers.get("content-type"),
+      "text/html; charset=utf-8",
+    );
+    assertEquals(response.status, 404);
+  });
+
+  await test.step("[get] /signin", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:94]:
```suggestion
  await test.step("GET /signin", async () => {
```

File: tests/http_test.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    assertFalse(response.ok);
+    assertFalse(response.body);
+    assertStringIncludes(
+      response.headers.get("location")!,
+      "https://github.com/login/oauth/authorize",
+    );
+    assertEquals(response.status, 302);
+  });
+
+  await test.step("[get] /signout", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:109]:
```suggestion
  await test.step("GET /signout", async () => {
```

File: tests/http_test.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      CONN_INFO,
+    );
+
+    assertFalse(response.ok);
+    assertFalse(response.body);
+    assertEquals(response.headers.get("location"), "/");
+    assertEquals(response.status, 302);
+  });
+
+  await test.step("[get] /stats", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:121]:
```suggestion
  await test.step("GET /stats", async () => {
```

File: tests/http_test.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    assert(response.ok);
+    assertInstanceOf(response.body, ReadableStream);
+    assertEquals(
+      response.headers.get("content-type"),
+      "text/html; charset=utf-8",
+    );
+    assertEquals(response.status, 200);
+  });
+
+  await test.step("[get] /submit", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:136]:
```suggestion
  await test.step("GET /submit", async () => {
```

File: tests/http_test.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    assertFalse(response.ok);
+    assertFalse(response.body);
+    assertEquals(
+      response.headers.get("location"),
+      "/signin?from=http://localhost/submit",
+    );
+    assertEquals(response.status, 303);
+  });
+
+  await test.step("[post] /submit", async () => {
```
Comment (iuioiua) [inline tests/http_test.ts:151]:
```suggestion
  await test.step("POST /submit", async () => {
```

**Review by iuioiua — APPROVED** [review]

**Comment by iuioiua** [comment]
> > Let's ignore test coverage for `/routes`, `/components` and `/islands`. See https://docs.codecov.com/docs/ignoring-paths
> 
> added the `ignore` section as suggested.
> 
> didn't came to mind to check the coverage, yet it makes sense that we took this big hit. due to `manifest` we'd import everything into the newly created http test set, correct?
> 
> as for the now ignored dirs it might make sense to introduce e2e tests at some point, wdyt?

Yep! In fact, what if we move the newly created e2e tests from this PR to the root of the folder and name it `e2e_test.ts`?

**Review by iuioiua — APPROVED** [review]
LGTM! Thanks for this!

**Pull request merged**